Add/improve errors for `.localAccess()` calls to non-local elements

bradcray opened the pull request:
This PR adds and improves errors for `.localAccess()` calls that access remote elements.

The original motivation for the PR was to fix the case noted in issue #25747 in which `.localAccess()` on a default rectangular array would always pass, whether that array was stored locally or on a remote locale (captured in `test/arrays/locality/localAccess/localAccess.chpl`).  It also improves the error message for `.localAccess()` violations on block arrays by expressing the error as a locality error rather than a simple out-of-bounds error, as before (also reported in #25747, captured in `test/arrays/locality/localAccess/localAccess-block*.chpl`).  These new tests make sure that these errors are generated for read and write cases as well as for POD vs. non-POD arrays, since our modules have distinct dsiAccess overloads for these cases.

The check itself is implemented as an additional/optional check in `checkAccess()` which is called for all array accesses when checking is on.  It now takes an optional `ensureLocal` param that is false by default, but set to true for all `.localAccess()` code paths to enable the bounds checking.  The logic itself generates different error messages depending on whether the array is completely local to a single (remote) locale, distributed but the current locale owns nothing, or distributed but the index is outside the current locale's bounds.

The implementation approach I originally pursued to determine whether an access was in-bounds was to use the array's `localSubdomains()` call to check that the localAccess's indices were in-bounds.  However, this caused problems for the Stencil distribution since its localSubdomain(s) (correctly) only refer to the indices that the locale truly owns, but not ones that are part of its "fluff" (cached ghost cells).  Yet, we also want that fluff to be accessible using localAccess() since that's how we get performance when the compiler can't optimize them automatically, requiring a slightly more sophisticated approach described in the next paragraph.  New test `test/distributions/stencil/localAccess.chpl` locks in a very simple case of this Stencil distribution pattern, with some accesses passing due to true ownership, some due to fluff ownership, and one generating an error due to a remote access.

The Stencil distribution challenge led to the approach taken here, which is to introduce a new internal iterator called `chpl__localStoredSubdomains` whose purpose is to yield a locale's subdomains, including indices that it doesn't truly own, like the fluff of the stencil distribution.  This is implemented using an optional call on the array, `doiLocalStoredSubdomains()`, and if not supported, we fall back to using the `localSubdomains()` call instead.  I then implemented `doiLocalStoredSubdomains()` on Stencil distributions.

Along the way, I found that `dsiLocalSubdomains()` calls weren't implemented for Block, Cyclic, and Stencil, so added those as well.

Then, while testing this, I found that our current tests that lock in support for using oversubscribed block/cyclic/stencil distributions (that is, ones in which a single locale appears in targetLocales multiple times) were getting away with using `.localAccess()` to access elements that weren't actually local, so I updated them to only update local array elements.  The reason for this is that the `dsiAccess()` routine on these distributions simply falls back to a normal `dsiAccess()` in the event of oversubscription—meaning that we're also losing some performance there in addition to not actually doing a local access...  I also found that the stencil variant of this test was never set up to run with multiple locales, so updated it to do so.

Note that much of the code in this effort may not be optimal performance-wise (e.g., iterating over local subdomains when the common case is that a locale only has a single subdomain), but that this seemed acceptable—at least for the time being—since this code is only called when checks are on.

Two bad files `test/expressions/if-expr/*.bad` had to be updated due to shifting IDs in the AST…  :(

Resolves #25747.

bradcray commented:
@e-kayrakli : I hate to say it, but I think I'm seeing another mis-application of the auto-local-access optimization (possibly equivalent to the previous one… it's Friday afternoon, so I'm thinking lazily) in `test/multilocale/diten/localBlock/needMultiLocales/localBlock5.chpl`.  Ironically, it's a case when .localAccess is applied to a remote DR array similar to #25747 that you filed.  Or maybe that's not ironic… maybe in looking into the previous issue, you came across and filed this?  Again, Friday, so I'm thinking lazily.

e-kayrakli commented:
I'll take a look. But looking at the code:

```chpl
var A: [1..5] shared C?;

on Locales(1) {
  [i in A.domain] A(i) = new shared C(i);
}
```

I want this `forall` to move back to `Locales[0]`, and would have guessed that it does that already until couple of months ago. This is related to https://github.com/chapel-lang/chapel/issues/25351.

Alas, I know that's not the case today and that ALA needs to be fixed. I had created https://github.com/chapel-lang/chapel/pull/25753 in a rushed way before taking a break. I'll revisit that PR soon and can probably lump the fix for that case in there.

e-kayrakli commented:
@bradcray the current solution I want to proceed with is not ideal, so I wanted to make sure you are aware of it.

The crux of the problem is that `[i in A.domain] ... A(i) ...` is always local with distributed arrays. But it may not be for DR. This will always require a dynamic check for DR (we can eliminate some by checking if there's an `on` between the def of `A` and this loop, but that's a lot of work and nearly impossible to perfect statically). Currently there's nothing laid out in support for array/domain/dists that can support ALA through only such dynamic checks. This makes it a lot of work, especially considering ALA doesn't make much sense for DR arrays in any case.

Where this could potentially hurt is auto-aggregation (AA) which builds on top of ALA's analysis. Consider the following case based on a more complicated AA unit test, which I believe was in turn based on some Arkouda code:

```chpl
  var destArr: [dom] int;  // assume `dom` is DR
  var sourceArr: ... something potentially distributed or remote ...

  // ... a lot of code, probably on statements ...

  forall i in dom with (ref destArr) {
    destArr[i] = sourceArr[generateIdx(i)]; // RHS will be seen as non-local
  }
```

On `main`, we aggregate this. But with the realization in this discussion, this is actually not aggregatable.

The solution I am proposing here is to disable ALA (and by connection AA) for DRs for the time being. There's already a flag wired in for that, I'll just have to switch it to `false` by default. A gasnet paratest doesn't show any other issues.

bradcray commented:
> The solution I am proposing here is to disable ALA (and by connection AA) for DRs for the time being.

Thanks for the update.  I'm completely fine with that approach, potentially even for the long-term.

e-kayrakli commented:
I just merged https://github.com/chapel-lang/chapel/pull/25840. @bradcray that should have solved the issue you were facing with this branch. Let me know if you don't see it fixed on `main`.

bradcray commented:
Thanks!  I'll try to get back to this next week.

bradcray commented:
@e-kayrakli : I confirmed that your PR fixed the issues related to the DR cases.  Unfortunately, one thing I didn't anticipate is that my general strategy of using localSubdomain() calls doesn't work properly for StencilDist because that returns the elements that the locale logically "owns", but not the ghost cells, so I'll need to do something different/special for it.

bradcray commented:
Think I got Stencil distributions working as expected now.  Will have to wait until my next session to see how a paratest works.

e-kayrakli commented on modules/dists/BlockDist.chpl:
```diff
@@ -1953,6 +1959,17 @@ proc BlockDom.dsiLocalSubdomain(loc: locale) {
+iter BlockDom.dsiLocalSubdomains(loc: locale) {
+  for locid in dist.chpl__locToLocIdxs(loc) {
+    var inds = chpl__computeBlock(locid, dist.targetLocDom, dist.boundingBox, dist.boundingBox.dims());
+    yield whole[(...inds)];
```
I am surprised by this implementation. Shouldn't it be just `yield this.locDoms[locid].myBlock;`?

bradcray replied:
Hmm, good question.  I took this approach due to the `else` clause in the `dsiLocalSubdomain()` routine just above, and to make it work with oversubscription; but in retrospect, I think you're right that I over-engineered it.  Specifically, I believe I was thinking that `myBlock` would only work in a non-oversubscribed environment, but with the loop over `locid`s, I think you're probably right.  I'll check that out to make sure that it works and doesn't introduce communication.

bradcray replied:
Ooh, and I can update the tests to do the localAcesss calls within `local` blocks as a way to both verify this and lock it in going forward.

bradcray replied:
> But then in my final review, I realized that the locale argument was never used / always defaulted to here, so removed that logic. So again, I think you're right.

Ah wait, it is necessary (if regrettable from a complexity standpoint) to do here since this routine takes a locale argument, so may be answering local subdomain queries about a remote locale's ownership, and we don't want to communicate with that remote locale to answer the question.

We could simplify / streamline the (common) `locale == here` case by either putting a conditional into this loop to check for it, or by creating two overloads of the routine, one which takes no argument (defaulting to `here`) and the other which doesn't—then updating ChapelArray.chpl's code to similarly split the implementation into two versions, each of which calls the corresponding one here.  I don't feel strongly inclined to do either of those things in this PR because I worry about keeping the two implementations in sync and would need to add more testing to distinguish between the remote and local cases.  It also feels a little like a change to the public interface (though it's one that I think would be acceptable since it wouldn't break existing code); or we could keep the same public interface and put a conditional into the outer call, but that doesn't seem great either.

That said, the `StencilDom.doiLocalStoredSubdomains` case should not be as expensive as I was fearing in the comment you commented on now that we don't accept a locale argument in that iterator, so that comment does need to be updated (and I can add a locality check to that test to lock that in as well).

e-kayrakli replied:
👍 I see. I agree that the solution you outlined would be good, but shouldn't be a part of this PR.

e-kayrakli reviewed (commented):
Thanks for this PR! I have two points in the inline comments:

- Your preference of the `dsiLocalSubdomains` implementation feels less performant and maintainable than a simple lookup
- It'd be good to add a test for illegal localAccess in an oversubscribed distribution

bradcray commented on modules/dists/StencilDist.chpl:
```diff
@@ -2346,6 +2352,37 @@ proc StencilDom.dsiLocalSubdomain(loc: locale) {
+    // Could calculate the indices directly, as in the normal
+    // subdomains case above, but that's more complicated for
+    // StencilDist, and this isn't on performance-critical paths
+    // anyway, so I'm just looking it up.
```
Ah, there's one other piece of history here that I was forgetting, but that this comment reminds me of.  I originally wrote these iterators to take a `locale` argument, and didn't want to do communication to answer the question.  But then in my final review, I realized that the locale argument was never used / always defaulted to `here`, so removed that logic.  So again, I think you're right.

bradcray replied:
I've now updated the comment to indicate that this is guaranteed to be local since we're only ever making this query for `here` (whereas the traditional localSubdomains() queries can be called with any locale argument to determine what subdomains are local to a given locale).

bradcray commented on test/distributions/block/duplicateTargetLocales.allow.good:
```diff
@@ -1,3 +1,3 @@
 1 0 0 0 0 0 0 0 0 10
-1 0 0 0 0 0 0 0 0 10
+1 0 0 0 0 0 7 0 0 0
```
Yeah, you're correct.  We were permitting illegal localAccesses in oversubscribed cases (and still do with checks off), and you're also correct that adding a test of that would be good.  Perhaps I can just add it to the end of these tests.

This also makes me realize: I'll need to skipif these tests with `--fast` and `--no-checks` compilations, right?

e-kayrakli replied:
Or, just stick a `--checks` in their compopts, which feels a bit easier and somewhat better as the nature of this test will now become also about testing the `--checks` behavior. I don't have any strong opinions here, though.

bradcray replied:
I think I've just verified that compopts is sufficient…

e-kayrakli replied:
:+1:

bradcray commented:
@e-kayrakli : OK, I think this is ready for another look.  Apologies for all the talking to myself in comments and backpedaling—I am sometimes too easily convinced I'm wrong.  To recap, since your original review:

* I haven't changed the implementations of the dsiLocalSubdomains() calls because the current implementations are communication-free when called with any locale, which is a property I think we want to preserve.  With more effort, we could use different implementations for when the argument locale is `==here` vs. `!= here`, but I don't think that additional complexity (and change) is worth the effort, at least as part of this PR.
* I have updated my tests to do the localAccess() calls within `local` blocks to ensure that the current calls are not doing communication, and to lock this in going forward.  Note that I have not tested making a dsiLocalSubdomains call from a remote locale to make sure communication is not required (but could, if desired)
* I updated the out-of-date comment in the new doi* iterator on Stencil now that it doesn't take an arbitrary locale argument anymore; the current approach is communication-free as a result, so not at all a compromise
* I added the missing checks you noted that lock in that we correctly get locality errors now in the oversubscribed case for a non-local localAccess() call where we were not before this PR; I did this as updates to the duplicateTargetLocales tests; I also added new comm-none .good files for these cases since they don't have the OOB issue
* I also remembered during this process that (earlier) I'd discovered that the stencil version of duplicateTargetLocales.chpl wasn't actually running using multiple locales, but then forgot to act on that at all.  So here, I've updated it to handle multiple locales, and in the same ways that I updated the block/cyclic cases.
* I added the `--checks` flags we discussed to compopts files to try and head off failures in `--fast` testing
